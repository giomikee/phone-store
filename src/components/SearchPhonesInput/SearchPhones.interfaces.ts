export interface Props {
    onSearchLoading: (isLoading: boolean) => void;
}