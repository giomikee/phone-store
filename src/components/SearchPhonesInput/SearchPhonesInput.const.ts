export const START_SEARCH_DELAY = 1000;