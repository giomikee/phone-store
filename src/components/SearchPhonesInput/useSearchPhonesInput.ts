import { useDispatch } from 'react-redux';
import { useNavigate } from 'react-router-dom';
import { useEffect, useRef, useState } from 'react';
import { setAreSearchResultsDisplayed, setSearchedPhones } from '../../state/phone/phone.slice';
import { getPhones } from '../../services';
import { PAGES } from '../../constants';
import { START_SEARCH_DELAY } from './SearchPhonesInput.const';
import { useStoreSelectors } from '../../state/useStoreSelectors';

export function useSearchPhonesInput(onSearchLoading: (isLoading: boolean) => void) {
    const navigate = useNavigate();
    const dispatch = useDispatch();

    const { searchedPhones, areSearchResultsDisplayed } = useStoreSelectors();

    const isFirstRender = useRef(true);
    const [isLoading, setIsLoading] = useState(false);
    const [keywords, setKeywords] = useState('');
    const timeoutIdRef = useRef<NodeJS.Timeout | undefined>(undefined);
    const resultsText = areSearchResultsDisplayed ? `(${searchedPhones.length} resultados)` : '';

    const startSearch = () => {
        setIsLoading(true);
        onSearchLoading(true);
        clearTimeout(timeoutIdRef.current);
    };

    const clearSearchResults = () => {
        dispatch(setSearchedPhones([]));
        setIsLoading(false);
        onSearchLoading(false);
        dispatch(setAreSearchResultsDisplayed(false));
    };

    useEffect(() => {
        if (isFirstRender.current) {
            isFirstRender.current = false;
            return;
        }

        startSearch();

        if (!keywords) {
            clearSearchResults();

            return;
        }

        timeoutIdRef.current = setTimeout(async () => {
            try {
                const searchedPhones = await getPhones({ search: keywords });
                dispatch(setSearchedPhones(searchedPhones));
            } catch (error) {
                console.error(error);
                navigate(PAGES.error);
            } finally {
                setIsLoading(false);
                onSearchLoading(false);
                dispatch(setAreSearchResultsDisplayed(true));
            }
        }, START_SEARCH_DELAY);
    }, [keywords]);

    return {
        keywords,
        isLoading,
        isFirstRender,
        resultsText,
        timeoutIdRef,
        clearSearchResults,
        setKeywords,
        startSearch,
    };
}