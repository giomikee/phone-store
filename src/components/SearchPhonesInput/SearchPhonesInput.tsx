import { FormHelperText, IconButton, TextField } from '@mui/material';
import SearchIcon from '@mui/icons-material/Search';
import HourglassTopIcon from '@mui/icons-material/HourglassTop';
import BackspaceIcon from '@mui/icons-material/Backspace';
import type { Props } from './SearchPhones.interfaces';
import { useSearchPhonesInput } from './useSearchPhonesInput';

const SearchPhonesInput = ({onSearchLoading}: Props) => {
    const {
        keywords,
        setKeywords,
        isLoading,
        resultsText,
    }= useSearchPhonesInput(onSearchLoading)

    return (
        <TextField
            fullWidth
            label={`Buscar MÃ³viles ${resultsText}`}
            id="search-phones-input"
            data-testid="search-phones-input"
            name="search-phones-input"
            slotProps={{
                input: {
                    endAdornment: <>
                        {
                            keywords.length > 0 &&
                                <IconButton 
                                    aria-label='borrar busqueda'
                                    data-testid='clear-search-button'
                                    onClick={() => setKeywords('')}>
                                    <BackspaceIcon />
                                </IconButton>
                        }
                        {isLoading ? <HourglassTopIcon color='disabled' /> : <SearchIcon color='disabled' />}
                    </>,
                },
                formHelperText: <FormHelperText>Weight</FormHelperText>
            }}
            value={keywords}
            onChange={e => setKeywords(e.target.value)}
        />
    )
}

export default SearchPhonesInput