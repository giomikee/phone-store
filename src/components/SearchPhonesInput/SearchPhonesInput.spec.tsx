import { beforeAll, describe, expect, it, vi, type Mock } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import '@testing-library/jest-dom/vitest';
import SearchPhonesInput from './SearchPhonesInput';
import { useSearchPhonesInput } from './useSearchPhonesInput';

vi.mock('./useSearchPhonesInput', () => ({
    useSearchPhonesInput: vi.fn(),
}));

describe('Given SearchPhonesInput component', () => {
    const defaultProps = {
        onSearchLoading: vi.fn(),
    };
    const mockedUseSearchPhonesInput = {
        keywords: 'keywords',
        resultsText: '',
        isLoading: false,
        setKeywords: vi.fn(),
    };

    beforeAll(() => {
        (useSearchPhonesInput as Mock).mockReturnValue(mockedUseSearchPhonesInput);
    });

    describe('When the component is rendered', () => {
        it('Then it should match the snapshot', () => {
            const { container } = render(<SearchPhonesInput {...defaultProps} />);
            expect(container).toMatchSnapshot();
        });
    });

    describe('When search phones input is updated with "foobar"', () => { 
        it('Then it should call setKeywords with "foobar"', () => {

            render(<SearchPhonesInput {...defaultProps} />);

            const input = screen.getByTestId('search-phones-input').querySelector('[name="search-phones-input"]');

            expect(input).not.toBeNull();

            fireEvent.change(input!, { target: { value: 'foobar' } });

            expect(mockedUseSearchPhonesInput.setKeywords).toHaveBeenCalledWith('foobar');
        });
    });

    describe('When clear search button is clicked', () => { 
        it('Then it should call setKeywords with an empty string', () => {
            render(<SearchPhonesInput {...defaultProps} />);

            const clearButton = screen.getByTestId('clear-search-button');

            fireEvent.click(clearButton);

            expect(mockedUseSearchPhonesInput.setKeywords).toHaveBeenCalledWith('');
        });
    });
});