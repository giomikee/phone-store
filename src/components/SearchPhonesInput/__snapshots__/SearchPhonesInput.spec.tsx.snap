// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given SearchPhonesInput component > When the component is rendered > Then it should match the snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-cmpglg-MuiFormControl-root-MuiTextField-root"
    data-testid="search-phones-input"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-113d811-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for="search-phones-input"
      id="search-phones-input-label"
    >
      Buscar Móviles 
    </label>
    <div
      class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-adornedEnd css-1n04w30-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-1dune0f-MuiInputBase-input-MuiOutlinedInput-input"
        id="search-phones-input"
        name="search-phones-input"
        type="text"
        value="keywords"
      />
      <button
        aria-label="borrar busqueda"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-53g0n7-MuiButtonBase-root-MuiIconButton-root"
        data-testid="clear-search-button"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1umw9bq-MuiSvgIcon-root"
          data-testid="BackspaceIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M22 3H7c-.69 0-1.23.35-1.59.88L0 12l5.41 8.11c.36.53.9.89 1.59.89h15c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2m-3 12.59L17.59 17 14 13.41 10.41 17 9 15.59 12.59 12 9 8.41 10.41 7 14 10.59 17.59 7 19 8.41 15.41 12z"
          />
        </svg>
      </button>
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeMedium css-3nihpm-MuiSvgIcon-root"
        data-testid="SearchIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
        />
      </svg>
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-18p5xg2-MuiNotchedOutlined-root-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-ex8a5f-MuiNotchedOutlined-root"
        >
          <span>
            Buscar Móviles 
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
